docs/weave-tutorial/06-buckets.md

```diff
@@ -0,0 +1,131 @@
+---
+id: buckets
+title: Buckets
+sidebar_label: Buckets
+---
+
+> [PR#8](https://github.com/iov-one/tutorial/pull/8): _Create buckets_
+\
+[PR#9](https://github.com/iov-one/tutorial/pull/9): _Add indexer for market using marketID, askTicker, bidTicker_
+
+When running your handlers, you get access to the root **KVStore**, which is an abstraction level similar to boltdb or leveldb. It is our data warehouse. An extension can opt-in to using one or more Buckets to store the data. Buckets offer the following advantages:
+
+- Isolation between extensions (each Bucket has a unique prefix that is transparently prepended to the keys)
+- Type safety (enforce all data stored in a Bucket is the same type, to avoid parse errors later on)
+- Indexes (Buckets are well integrated with the secondary indexes and keep them in sync every time data is modified)
+- Querying (Buckets can easily register query handlers including prefix queries and secondary index queries)
+
+All extensions from weave use Buckets, so for compatibility as well as the features, **please use Buckets in your app**, unless you have a very good reason not to (and know what you are doing).
+
+To do so, you will have to wrap your state data structures into [Objects](https://godoc.org/github.com/iov-one/weave/orm#Object). The simplest way is to use `SimpleObj`:
+
+```go
+// It can be used as a template for type-safe objects
+type SimpleObj struct {
+    key   []byte
+    value CloneableData
+```
+
+And extend your protobuf objects to implement `CloneableData`:
+
+```go
+    Clone() Object
+}
+
+// CloneableData is an intelligent Value that can be embedded
+// in a simple object to handle much of the details.
+```
+
+This basically consists of adding _Copy()_ and _Validate()_ to the objects in `codec.pb.go`. On [Models](weave-tutorial/04-models.md) section, we implemented _Copy()_ and _Validate()_ as you remember. Now it makes sense right!
+
+## Dive into Code
+
+Let's define `MarketBucket` that will hold `Market` information and write a function that creates a market. This is a basic `morm/model_bucket` without any secondary index.
+
+```go
+type MarketBucket struct {
+    morm.ModelBucket
```

Comment:
I think we should not encourage to use inheritance. This construct was needed when using `orm.Bucket` because of limited functionality it provides. Both `orm.ModelBucket` and `morm.ModelBucket` can be used as they are. When used in our extensions, we do not ever create a separate bucket type ([1](https://github.com/iov-one/weave/blob/master/x/distribution/model.go#L100), [2](https://github.com/iov-one/weave/blob/master/x/escrow/model.go#L89), [3](https://github.com/iov-one/weave/blob/master/x/distribution/model.go#L100)) and return `orm.ModelBucket` interface directly.

This is my preference and maybe others should share their opinion here as well.

Reply:
If this is the code standard you follow, we should use the same way in the tutorial to keep the standard. @alpe @ruseinov what do you say?